practicas/FredinVazquez/P1/P1_Niveles_del_lenguaje.py

```diff
@@ -0,0 +1,961 @@
+# 
+# Entre estos para una implementación sencilla se podría implementar estas dos soluciones con jaccard y distancia de edición.
+# 
+# Ambas soluciones resultan viables, me recuerdan a un problema de programación competitiva sobre *edit distance*, basicamente es un problema para introducir el algoritmo de distancia de Levenshtein. Basado en la idea de que el algoritmo nos devuelve la cantidad de mínima de operaciones a realizar para transformar la palabra en la otra, podemos poner un umbral para saber cuánto es lo permitido para considerar que ambas palabras son suficientemente similares para poder usar su transcripción fonológica como una aproximación a la palabra de la cual no conocemos.
+# 
+# A continuación se muestra la implementación de edit_distance, sin embargo ya existe una versión implementada más optimizada en nltk.
+# 
+# ---
+
+# In[6]:
+
+
+palabra1 = 'hervívoros'
+palabra2 = 'carnívoro'
+
+# Similitud de jaccard
+def jaccard_similarity_ngrams(word1, word2, n=2):
+  '''
+  Similitud de jaccard usando ngramas para obtener un grado de similitud entre
+  dos palabras.
+
+  Input:
+    word1: palabra 1 para hacer la comparación
+    word2: palabra 2 para hacer la comparación
+    n: cantidad de n-gramas
+
+  Output:
+    Porcentaje de la similitud de las dos palabras de entrada.
+  '''
+  set1, set2 = set(ngrams(word1, n)), set(ngrams(word2, n))
+  return 1 - jaccard_distance(set1, set2)
+
+
+# Edit Distance
+def min_edit_distance(source, target, ins_cost = 1, del_cost = 1, rep_cost = 2):
+    '''
+    Distancia de Levenshtein para obtener la cantidad de cambios necesarios
+    para que el source sea el mismo que el target.
+
+    Input:
+      source: una cadena de texto que corresponde a la cadena de inicio.
+      target: una cadena de texto que corresponde a la cadena de destino.
+      ins_cost: un número entero que establece el costo de inserción.
+      del_cost: un número entero que establece el costo de eliminación.
+      rep_cost: un número entero que establece el costo de reemplazo.
+    Output:
+      D: una matriz de tamaño (len(source) + 1) × (len(target) + 1) que contiene las distancias mínimas de edición.
+      med: la distancia mínima de edición (med) necesaria para convertir la cadena de inicio en la cadena de destino.
+    '''
+    m = len(source)
+    n = len(target)
+
+    D = np.zeros((m+1, n+1), dtype=int)
+
+    for row in range(1,m+1):
+        D[row,0] = row
+
+    for col in range(1,n+1):
+        D[0,col] = col
+
+    for row in range(1,m+1):
+        for col in range(1,n+1):
+            r_cost = rep_cost
+
+            if source[row-1] == target[col-1]:
+                r_cost = 0
+
+            D[row,col] = min(D[row-1,col] + del_cost, min(D[row,col-1] + ins_cost, D[row-1,col-1] + r_cost)) # Ecuación de recurrencia para obtener el costo mínimo de las tres operaciones
```

Comment:
Para líneas muy largas creo que vale la pena formatearlas y los coments se acostumbra que vayan una línea antes para que sean ilustrativos

```suggestion
            # Ecuación de recurrencia para obtener el costo mínimo de las tres operaciones
            D[row, col] = min(
                D[row - 1, col] + del_cost,
                min(D[row, col - 1] + ins_cost, D[row - 1, col - 1] + r_cost),
            ) 
```